### ISSUE TITLE
Unexpected type error

### ISSUE THREAD

**Issue by altomani** [microsoft/pyright#2895]
In the following code I get an error 
```
Argument of type "dict[str, MyFunc[Unknown, V@MyFunc]]" cannot be assigned to parameter "__iterable" of type "Iterable[tuple[_KT@defaultdict, _VT@defaultdict]]" in function "__init__"
```
on `{"x": MyFunc(lambda f: f("a"))})`. 

```python
from collections import defaultdict
from typing import Callable, Generic, Mapping, Optional, TypeVar


K = TypeVar('K')
V = TypeVar('V')

MyFuncType = Callable[ [Callable[[K], V]], V]

class MyFunc(Generic[K, V]): 
    def __init__(self, g: MyFuncType[K,V]) -> None:
        self.g = g

MyFuncMapping = Mapping[K, Optional[MyFunc[K, V]]]

# Type error in PyRight
my_func_defaultdict: MyFuncMapping[str, int] = defaultdict(lambda: None,  {"x": MyFunc(lambda f: f("a"))})
```

The strange thing is that, in the same context, the two following declarations do not give any error:
```python
# No type error if my_func_x declared separatey
my_func_x: MyFunc[str,int] = MyFunc(lambda f: f("a"))
my_func_defaultdict: MyFuncMapping[str, int] = defaultdict(lambda: None,  {"x": my_func_x})

# No type error if dict instead of defaultdict
my_func_dict: MyFuncMapping[str, int] = {"x": MyFunc(lambda f: f("a"))}
my_func_dict2: MyFuncMapping[str, int] = dict({"x": MyFunc(lambda f: f("a"))})
```

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.
